### PR TITLE
Add agentic trajectory snapshot metrics

### LINKED ISSUE #669: OpenSearch-VL alignment: add agentic trajectory dataset contracts - Milestone 2: materialize and validate trajectory packages - Unit 1: Extend training dataset resolution to emit normalized trajectory snapshots.
## Executable Unit

Extend training dataset resolution to emit normalized trajectory snapshots.

Parent direction: #664

Parent milestone: #668

Governing spec:

- `docs/agentic-trajectory-dataset-contract.md`

## File Scope

- `services/mlx-worker-python/worker/model_ops/training_dataset.py`
- LoRA preparation paths that call `write_normalized_dataset_snapshot(...)`
- `services/mlx-worker-python/tests/test_training_dataset_builder.py`
- LoRA model-ops tests if job preparation behavior changes
- Swift bridge files only if public CLI/API payloads change

## Required Tests

- Snapshot manifest preserves `format: agentic_tool_trace`.
- Snapshot manifest records source dataset id, package path, version, split, train count, validation count, and quality metrics.
- Snapshot JSONL preserves normalized trajectory rows.
- LoRA or rollout preparation reads the normalized job snapshot after job start.
- Existing non-agentic dataset formats preserve current behavior.
- Changed-scope coverage for touched Python or Swift files is at least 95 percent or a precise gap is reported.

## Required Metrics

- snapshot write duration
- train sample count
- validation sample count
- trajectory trace digest
- cache hit/miss if materialization is cached
- `agentic_trace_count`
- `tool_call_count`
- `tool_observation_count`

## Known Gaps

- Does not add adapter or evaluation provenance fields.
- Does not implement new optimizer or model-training logic.
- Does not prove benchmark or evaluation performance.

## Acceptance

- Downstream consumers can rely on a job-local normalized trajectory snapshot.
- Snapshot provenance is stable enough for #672 and #673 artifact fields.

### LINKED ISSUE #670: OpenSearch-VL alignment: add agentic trajectory dataset contracts - Milestone 2: materialize and validate trajectory packages - Unit 2: Add validation metrics for trace length, tool-call count, media refs, reward coverage, and fatal-stage coverage.
## Executable Unit

Add validation metrics for trace length, tool-call count, media refs, reward coverage, and fatal-stage coverage.

Parent direction: #664

Parent milestone: #668

Governing spec:

- `docs/agentic-trajectory-dataset-contract.md`

## File Scope

- `services/mlx-worker-python/worker/model_ops/training_dataset.py`
- Any report/export helpers that surface dataset quality
- `services/mlx-worker-python/tests/test_training_dataset_builder.py`
- Focused report tests if quality metrics are projected into reports

## Required Tests

- Quality metrics include required trajectory counts for `agentic_tool_trace`.
- Turn count min, max, and average are deterministic.
- Media ref count is aggregated.
- Reward coverage count is aggregated.
- Fatal-stage field coverage and non-empty fatal count are distinct.
- Leakage count remains reported separately from fatal count.
- Non-agentic dataset quality metrics preserve current behavior.
- Changed-scope coverage for touched Python files is at least 95 percent or a precise gap is reported.

## Required Metrics

- `agentic_trace_count`
- `trace_turn_count_min`
- `trace_turn_count_max`
- `trace_turn_count_avg`
- `tool_call_count`
- `tool_observation_count`
- `media_ref_count`
- `reward_coverage_count`
- `fatal_stage_coverage_count`
- `fatal_trace_count`
- `leakage_count`

## Known Gaps

- Metrics describe package quality, not model quality.
- Does not add persisted benchmark or evaluation result artifacts.
- Does not implement semantic leakage detection.

## Acceptance

- Validation metrics are machine-readable and available before downstream consumption.
- Required fields from the governing spec are covered by focused tests.

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Adds `agentic_tool_trace` normalized snapshot provenance for source package, source dataset, split, schema, toolset, reward, leakage, and registry metadata.
- Adds trajectory quality metrics and a deterministic trace digest so LoRA SFT, RL alignment, benchmark, and evaluation consumers can verify package identity and coverage from the same snapshot contract.
- Preserves non-agentic snapshot manifests and keeps validation sample-limit loading on the existing fast path.

Closes #669
Closes #670

## Plan or Spec
- `docs/agentic-trajectory-dataset-contract.md`
- `docs/plans/2026-05-19-opensearch-vl-trajectory-snapshots.md`

## Commands Run
```text
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_training_dataset_builder.py::test_write_normalized_dataset_snapshot_applies_manifest_overrides services/mlx-worker-python/tests/test_training_dataset_builder.py::test_load_training_dataset_package_limits_validation_samples services/mlx-worker-python/tests/test_training_dataset_builder.py::test_load_training_dataset_package_stops_reading_validation_after_sample_limit
# 3 passed in 0.03s

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" COVERAGE_FILE=/tmp/opensearch-vl-trajectory-snapshots-after-merge.coverage uv run --project services/mlx-worker-python coverage run -m pytest -q services/mlx-worker-python/tests/test_training_dataset_builder.py
# 75 passed in 0.18s

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --project services/mlx-worker-python coverage json -o /tmp/opensearch-vl-trajectory-snapshots-after-merge-coverage.json
# Wrote JSON report

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --project services/mlx-worker-python python scripts/python_changed_line_coverage.py --coverage-json /tmp/opensearch-vl-trajectory-snapshots-after-merge-coverage.json --diff-from origin/main services/mlx-worker-python/worker/model_ops/training_dataset.py services/mlx-worker-python/tests/test_training_dataset_builder.py
# TOTAL 100.00% 67/67

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --project services/mlx-worker-python python scripts/pre_commit_gate.py via run_performance_report(root, changed_files, base_ref='origin/main')
# Status ok; 3 selected direct/gated probes; 0 regressions; 0 verification failures

make swift-test
# First full pre-commit attempt passed the macOS menubar package, then failed later on PR-scoped performance before the fast-path fix.
# Later full pre-commit attempts were blocked in this SSH/background session by macOS Pasteboard availability: make swift-test-menubar failed only in existing NSPasteboard copy assertions, and a minimal Swift NSPasteboard probe plus pbcopy also returned no writable pasteboard.
```

## Coverage and Metrics
- Changed-line coverage: `100.00%` (`67/67`) for `training_dataset.py` and `test_training_dataset_builder.py`.
- PR-scoped performance report: `.runtime/pre-commit-performance/20260519-122148-3ddd5136/report/report.md`.
- Direct/gated probes: `3`; status `ok`; regressions `0`; verification failures `0`.
- Quality/token summary: elapsed `-28.70%` improvement, peak bytes `+0.01%` neutral.
- Validation split partial selection: elapsed `+3.98%` neutral, peak bytes `-0.02%` neutral, validation count unchanged.
- Validation sample-limit loading: elapsed `+1.28%` neutral, peak bytes `+0.00%` neutral, validation sample count unchanged.
- Observability mode: evidence for snapshot manifest/quality metrics; probe overhead is covered by the PR-scoped performance report.
- Protocol changes: N/A, no protobuf schema changes.
- Dependency changes: N/A, no dependency or lockfile changes.

## Known Gaps
- This implements the M2 snapshot and validation-metric slice for #669 and #670 only.
- #672 and #673 remain for LoRA/RL adapter provenance and benchmark/evaluation provenance.
- No model-quality or evaluation-score improvement is claimed here.
- Full local pre-commit was not completed after the final fix because the current SSH/background macOS session has no writable `com.apple.pboard` service for existing menubar clipboard tests; CI should run those tests in a normal GUI-capable runner/session.

## Evidence Checklist
- [x] Relevant plan or spec is identified.
- [x] Behavior changes are reflected in the relevant docs.
- [x] Protocol changes include regenerated generated artifacts, or are N/A.
- [x] Dependency changes include updated lockfiles, or are N/A.
- [x] Relevant tests were run.
- [x] A metrics report is included, or `N/A` is stated explicitly with the reason.
- [x] Observability mode, probe overhead, and deferred evidence work are recorded.
- [x] Deferred work and known gaps are stated explicitly.

### CODE REVIEW

File: services/mlx-worker-python/worker/model_ops/training_dataset.py [hunk 1]
```diff
@@ -1856,6 +1860,7 @@ def _build_quality_and_token_stats(
         "dirty_samples": dirty_samples,
     }
     if format_name == "agentic_tool_trace":
+        agentic_metrics.pop("_trace_turn_count_sum", None)
         quality.update(agentic_metrics)
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/model_ops/training_dataset.py:1870]: Implemented in 6abc7932: `trace_turn_count_avg` is now finalized once after sample iteration, the per-sample recalculation was removed from `_update_agentic_trace_quality_metrics`, and the redundant manifest pop was removed. Re-ran focused tests, changed-line coverage (100%), and PR-scoped performance; all direct probes are ok.
